test/e2e/ownerreference_test.go

```diff
@@ -0,0 +1,282 @@
+	BeforeEach(func() {
+		createVsphereIdentitySecret(ctx, bootstrapClusterProxy)
+	})
+
+	capi_e2e.QuickStartSpec(ctx, func() capi_e2e.QuickStartSpecInput {
+		return capi_e2e.QuickStartSpecInput{
+			E2EConfig:             e2eConfig,
+			ClusterctlConfigPath:  clusterctlConfigPath,
+			BootstrapClusterProxy: bootstrapClusterProxy,
+			ArtifactFolder:        artifactFolder,
+			SkipCleanup:           skipCleanup,
+			Flavor:                pointer.String("ownerreferences"),
+			PostMachinesProvisioned: func(proxy framework.ClusterProxy, namespace, clusterName string) {
+				// Inject a client to use for checkClusterIdentitySecretOwnerRef
+				checkClusterIdentitySecretOwnerRef(ctx, proxy.GetClient())
+
+				// Set up a periodic patch to ensure the DeploymentZone is reconciled.
+				forcePeriodicDeploymentZoneReconcile(ctx, proxy.GetClient())
+
+				// This check ensures that owner references are resilient - i.e. correctly re-reconciled - when removed.
+				framework.ValidateOwnerReferencesResilience(ctx, proxy, namespace, clusterName,
+					framework.CoreOwnerReferenceAssertion,
+					framework.KubeadmBootstrapOwnerReferenceAssertions,
+					framework.KubeadmControlPlaneOwnerReferenceAssertions,
+					framework.ExpOwnerReferenceAssertions,
+					VSphereKubernetesReferenceAssertions,
+					VSphereReferenceAssertions,
+				)
+				// This check ensures that owner references are always updated to the most recent apiVersion.
+				framework.ValidateOwnerReferencesOnUpdate(ctx, proxy, namespace, clusterName,
+					framework.CoreOwnerReferenceAssertion,
+					framework.KubeadmBootstrapOwnerReferenceAssertions,
+					framework.KubeadmControlPlaneOwnerReferenceAssertions,
+					framework.ExpOwnerReferenceAssertions,
+					VSphereKubernetesReferenceAssertions,
+					VSphereReferenceAssertions,
+				)
+			},
+		}
+	})
+
+	// We must delete the secret created for the VSphereClusterIdentity and then delete it.
+	AfterEach(func() {
+		cleanupVSphereObjects(ctx, bootstrapClusterProxy)
+	})
+
+})
+
+var (
+	VSphereKubernetesReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+		// Need custom Kubernetes assertions for secrets. Secrets in the CAPV tests can also be owned by the vSphereCluster.
+		"Secret": func(owners []metav1.OwnerReference) error {
+			return framework.HasOneOfExactOwners(owners,
+				// Secrets for cluster certificates must be owned by the KubeadmControlPlane.
+				[]metav1.OwnerReference{kubeadmControlPlaneController},
+				// The bootstrap secret should be owned by a KubeadmConfig.
+				[]metav1.OwnerReference{kubeadmConfigController},
+				// Secrets created as a resource for a ClusterResourceSet can be owned by the ClusterResourceSet.
+				[]metav1.OwnerReference{clusterResourceSetOwner},
+				// Secrets created as an identityReference for a vSphereCluster should be owned but the vSphereCluster.
+				[]metav1.OwnerReference{vSphereClusterOwner},
```

Comment:
Probably mixing something up, but shouldn't in our case the secret be used via VSphereClusterIdentity and thus owned by the VSphereClusterIdentity?

Reply:
Okay I see the other case is covered below. Shouldn't this assertion then fail if it finds a secret owned by a VSphereClusterIdentity?

Ah or is it not looking in the capv-system ns?

Reply:
Independent of that, do you want to have this assertion given we don't have that case? (I'm fine, let's just maybe document it if that is the case)

Reply:
I'm fine with it as is - it makes the test work on Clusters which don't use ClusterIdentity if we wanted to introduce it for any of the quick start tests.

The secret in capv-system namespace is not handled in this function at all. The ownerGraph function doesn't take into account the `forceMove` labels so we can't add it to the graph easily. There's a completely separate function to check it here.

Reply:
Makes sense